src/pydvl/influence/frameworks/util.py

```diff
@@ -0,0 +1,284 @@
+    V is orthogonal
+    :param eigen_vals: diagonal of D
+    :param projections: the matrix V
+    """
+
+    eigen_vals: torch.Tensor
+    projections: torch.Tensor
+
+
+def lanzcos_low_rank_hessian_approx(
+    hessian_vp: Callable[[torch.Tensor], torch.Tensor],
+    matrix_shape: Tuple[int, int],
+    hessian_perturbation: float = 0.0,
+    rank_estimate: int = 10,
+    krylov_dimension: Optional[int] = None,
+    x0: Optional[torch.Tensor] = None,
+    tol: float = 1e-6,
+    max_iter: Optional[int] = None,
+    device: Optional[torch.device] = None,
+    eigen_computation_on_gpu: bool = False,
+) -> LowRankProductRepresentation:
+    """
+    Calculates a low-rank approximation of the Hessian matrix of the model's loss function using the implicitly
+    restarted Lanczos algorithm.
+
+
+    :param hessian_vp: A function that takes a vector and returns the product of the Hessian of the loss function
+    :param matrix_shape: The shape of the matrix, represented by hessian vector product.
+    :param hessian_perturbation: Optional regularization parameter added to the Hessian-vector product
+                                 for numerical stability.
+    :param rank_estimate: The number of eigenvalues and corresponding eigenvectors to compute.
+                          Represents the desired rank of the Hessian approximation.
+    :param krylov_dimension: The number of Krylov vectors to use for the Lanczos method.
+                             If not provided, it defaults to $min(model.num_parameters, max(2*rank_estimate + 1, 20))$.
+    :param x0: An optional initial vector to use in the Lanczos algorithm.
+               If not provided, a random initial vector is used.
+    :param tol: The stopping criteria for the Lanczos algorithm, which stops when the difference
+                in the approximated eigenvalue is less than `tol`. Defaults to 1e-6.
+    :param max_iter: The maximum number of iterations for the Lanczos method. If not provided, it defaults to
+                     $10*model.num_parameters$
+    :param device: The device to use for executing the hessian vector product.
+    :param eigen_computation_on_gpu: If True, tries to execute the eigen pair approximation on the provided
+                                     device via cupy implementation.
+                                     Make sure, that either your model is small enough or you use a
+                                     small rank_estimate to fit your device's memory.
+                                     If False, the eigen pair approximation is executed on the CPU by scipy wrapper to
+                                     ARPACK.
+    :return: A `LowRankProductRepresentation` instance that contains the top (up until rank_estimate) eigenvalues
+             and corresponding eigenvectors of the Hessian.
+    """
+
+    if eigen_computation_on_gpu:
+        try:
+            import cupy as cp
```

Comment:
We should add a new `extras_requires` for cupy. Maybe "gpu"? There is a nice idea on how to maintain this [here](https://hanxiao.io/2019/11/07/A-Better-Practice-for-Managing-extras-require-Dependencies-in-Python/)

Reply:
For now, I would stick with just adding it to extras_requires to keep it as simple as possible